### PR TITLE
feat(frontend): add shrink cast for condition.split_to_scan_ranges()

### PR DESCRIPTION
**Pull request by ZENOTME**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

* assign type suffix to smallint,int,bigint to solve #7900 

* add shrink cast which can cast a bigger range type to smaller range type, e.g.: bigint -> int 

related issue: #7916

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Comment by ZENOTME** [comment]
Do u think this implementation works? @xiangjinwu @chenzl25 PTAL 

TODO:
- support shrink cast in as_in_const_list
- support more type shrink cast

**Review by chenzl25 — APPROVED** [review]
LGTM. @xiangjinwu PTAL. I am not quite sure whether it is the best way to deal with the implicit cast and 'shrink' cast.

File: src/frontend/planner_test/tests/testdata/range_scan.yaml [hunk 1]
```diff
@@ -238,24 +238,28 @@
...
   before:
   - create_small
   sql: |
     SELECT * FROM mv WHERE x = 60000;
   batch_plan: |
-    BatchExchange { order: [], dist: Single }
-    └─BatchFilter { predicate: (mv.x = 60000:Int32) }
-      └─BatchScan { table: mv, columns: [mv.x], distribution: UpstreamHashShard(mv.x) }
-- name: When the constant has a larger type and cannot be casted to the column's type, we can't convert it to scan range yet.
+    BatchValues { rows: [] }
```
Comment (jon-chuang) [inline src/frontend/planner_test/tests/testdata/range_scan.yaml:247]: Will this be fixed?

File: src/frontend/src/expr/mod.rs [hunk 2]
```diff
@@ -176,18 +183,60 @@ impl ExprImpl {
...
+    fn cast_shrink_for_int(self, target: DataType) -> Result<CastResult> {
+        let (upper_bound, lowwer_bound) = match (self.return_type(), &target) {
+            (DataType::Int64, DataType::Int32) => (i32::MAX as i64, i32::MIN as i64),
+            (DataType::Int64, DataType::Int16) | (DataType::Int32, DataType::Int16) => {
+                (i16::MAX as i64, i16::MIN as i64)
+            }
+            _ => return Ok(CastResult::Failed),
+        };
+        if let Some(scalar) = self.eval_row_const()? {
+            let value = scalar.as_integral();
```
Comment (jon-chuang) [inline src/frontend/src/expr/mod.rs:215]: Hmm, will we be able to handle `float`, `decimal` as well?

File: src/frontend/src/expr/mod.rs [hunk 3]
```diff
@@ -176,18 +183,60 @@ impl ExprImpl {
...
+                Ok(CastResult::Success(self.cast_explicit(target)?))
+            }
+        } else {
+            Ok(CastResult::Failed)
+        }
+    }
+
     /// Shorthand to create cast expr to `target` type in implicit context.
-    pub fn cast_implicit(self, target: DataType) -> Result<ExprImpl> {
+    pub fn cast_implicit(self, target: DataType) -> std::result::Result<ExprImpl, ErrorCode> {
```
Comment (jon-chuang) [inline src/frontend/src/expr/mod.rs:229]: Hmm, why is it necessary to change the error type?

File: src/frontend/src/expr/mod.rs (hunk 2 above)
Comment (ZENOTME) [reply src/frontend/src/expr/mod.rs:215]: Yes

File: src/frontend/src/expr/mod.rs (hunk 3 above)
Comment (ZENOTME) [reply src/frontend/src/expr/mod.rs:229]: For RwError, it will capture the backtrace so the cost of constructing the RwError is high. There is a known issue #6131. 

In cast_implicit_or_shrink, We will store Error return form cast_implicit before call cast_shrink, so if cast_implicit return a RwError, the performance is unacceptable.

File: src/frontend/src/expr/mod.rs (hunk 2 above)
Comment (xiangjinwu) [reply src/frontend/src/expr/mod.rs:215]: Good question... So it is not enough to only check for minimum and maximum. For `where id < 2.5` and `where id > 2.5`, we need to round in different directions: `ceil` to `id < 3` and `floor` to `id > 2`. And when it is `id = 2.5`, this should evaluate to `false`.

There may be more complicated edge cases we are missing. Without optimization, the system always casts to a "larger" type and compare in that type. But in this optimization, we always want to compare in the type of order column, which may not be the "larger" type. Mimicking this `CMP_B(A_TO_B(x), A_TO_B(y))` via `CMP_A(B_TO_A(x), B_TO_A(y))` requires us to find the proper `B_TO_A` function given existing `A_TO_B`, `CMP_B` and `CMP_A` for all possible inputs.

More edge cases (PostgreSQL):
```
test=# select bigint '9223372036854775807' = double precision '9223372036854775806.5';
 ?column? 
----------
 t
(1 row)
test=# select (double precision '9223372036854775806.5')::bigint;
ERROR:  bigint out of range
test=# select (double precision '9223372036854775295')::bigint;
        int8         
---------------------
 9223372036854774784
(1 row)
```

Given it is an optimization, my suggestions is just to do it with best effort, and remain unoptimized for cases hard to handle correctly. int16/int32/int64 may be a good subset.

Also given its best-effort nature, the "cast_shrink" series helper functions should be limited to a mod for this special optimization, rather than on `ExprImpl`.

File: src/frontend/src/utils/condition.rs [hunk 4]
```diff
@@ -676,6 +754,139 @@ impl fmt::Debug for ConditionDisplay<'_> {
...
+            | (DataType::Int64, DataType::Int16)
+            | (DataType::Int32, DataType::Int16) => match shrink_int(const_expr, target)? {
+                ShrinkResult::OutUpperBound | ShrinkResult::OutLowerBound => Ok(None),
+                ShrinkResult::InRange(expr) => Ok(Some(expr)),
+            },
+            _ => Err(ErrorCode::BindError(format!(
+                "Cannot cast type \"{}\" to \"{}\".",
+                const_expr.return_type(),
+                target
+            ))
```
Comment (xiangjinwu) [inline src/frontend/src/utils/condition.rs:788]: This error is now thrown all the way up to the user:
```
dev=> create table t (v1 int primary key);
CREATE_TABLE
dev=> select * from t where v1 < 1.5;
ERROR:  QueryError: Bind error: Cannot cast type "numeric" to "integer".
```

But it should just be a signal to skip the scan range optimization. Talking about its caller `analyze_eq_const_expr` behavior, it should do one of the 3 things:
* return `Ok(true)` so that the upper layer `return Ok(false_cond())`
* assign `eq_conds` and `return Ok(false)`
* signal the upper layer it [cannot match a known optimization](https://github.com/risingwavelabs/risingwave/blob/88dc35e44554152a88366f6535d7a60b37a4bf9f/src/frontend/src/utils/condition.rs#L467) and it should just `other_conds.push(expr);`

We are lacking the last possibility right now.

**Comment by ZENOTME** [comment]
Summarize the todo work: 
- handle float, decimal and more type 
- there are some todo work in split_to_scan_range, like merge upperbound and lowerbound

For these todo work, I add the related plan test named `Cannot do sth` and I will fix them in later PR.

File: src/frontend/planner_test/tests/testdata/range_scan.yaml [hunk 5]
```diff
@@ -233,28 +233,94 @@
   batch_plan: |
     BatchExchange { order: [], dist: Single }
     └─BatchScan { table: orders_count_by_user_ordered, columns: [orders_count_by_user_ordered.user_id, orders_count_by_user_ordered.date, orders_count_by_user_ordered.orders_count], scan_ranges: [orders_count_by_user_ordered.orders_count = Int64(10)], distribution: UpstreamHashShard(orders_count_by_user_ordered.user_id, orders_count_by_user_ordered.date) }
+- name: Cannot push down for Mutiple lower bound 1.
+  before:
+  - create_table_and_mv_ordered
+  sql: |
+    SELECT * FROM orders_count_by_user_ordered WHERE user_id < 10 and user_id < 30
```
Comment (xiangjinwu) [inline src/frontend/planner_test/tests/testdata/range_scan.yaml:240]:
```suggestion
- name: Cannot push down for Mutiple upper bound 1.
  before:
  - create_table_and_mv_ordered
  sql: |
    SELECT * FROM orders_count_by_user_ordered WHERE user_id < 10 and user_id < 30
```
(same below)

**Review by xiangjinwu — COMMENTED** [review]
The current implementation looks correct, but I would like to try harder and see if the code structure, function names and doc comments can be easier to understand 🤯

File: src/frontend/planner_test/tests/testdata/range_scan.yaml (hunk 5 above)
Comment (ZENOTME) [reply src/frontend/planner_test/tests/testdata/range_scan.yaml:240]: Sorry, my fault...I misuse lower bound and upper bound.

File: src/frontend/src/utils/condition.rs [hunk 6]
```diff
@@ -676,6 +742,88 @@ impl fmt::Debug for ConditionDisplay<'_> {
...
+                ShrinkResult::InRange(expr) => Ok(ResultForCmp::Success(expr)),
+                ShrinkResult::OutUpperBound => Ok(ResultForCmp::OutUpperBound),
+                ShrinkResult::OutLowerBound => Ok(ResultForCmp::OutLowerBound),
+            },
+            _ => Err(()),
+        }
+    }
+
+    fn shrink_integral(const_expr: ExprImpl, target: DataType) -> Result<ShrinkResult, ()> {
+        let (upper_bound, lowwer_bound) = match (const_expr.return_type(), &target) {
```
Comment (xiangjinwu) [inline src/frontend/src/utils/condition.rs:800]: lower_bound

File: src/frontend/src/utils/condition.rs [hunk 7]
```diff
@@ -676,6 +742,88 @@ impl fmt::Debug for ConditionDisplay<'_> {
     }
 }
 
+mod execpetion_cast {
+    use risingwave_common::types::DataType;
+    use risingwave_pb::expr::expr_node::Type;
+
+    use crate::expr::{Expr, ExprImpl};
```
Comment (xiangjinwu) [inline src/frontend/src/utils/condition.rs:749]:
```suggestion

    use crate::expr::{Expr, ExprImpl, ExprType};
```

**Review by xiangjinwu — APPROVED** [review]
LGTM. Thanks for the refactor!
